src/reactor.rs

```diff
@@ -465,57 +478,109 @@ impl Source {
-pub struct Readable(Ready);
+pub struct Readable<'a, T>(Ready<&'a crate::Async<T>, T>);
+
+impl<T> Future for Readable<'_, T> {
+    type Output = io::Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ready!(Pin::new(&mut self.0).poll(cx))?;
+        log::trace!("readable: fd={}", self.0.handle.source.raw);
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<T> fmt::Debug for Readable<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Readable").finish()
+    }
+}
+
+/// Future for [`Async::readable_owned`](crate::Async::readable_owned).
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct OwnedReadable<T>(Ready<Arc<crate::Async<T>>, T>);
 
-impl Future for Readable {
+impl<T> Future for OwnedReadable<T> {
     type Output = io::Result<()>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         ready!(Pin::new(&mut self.0).poll(cx))?;
-        log::trace!("readable: fd={}", self.0.source.raw);
+        log::trace!("readable_owned: fd={}", self.0.handle.source.raw);
         Poll::Ready(Ok(()))
     }
 }
 
+impl<T> fmt::Debug for OwnedReadable<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OwnedReadable").finish()
+    }
+}
+
 /// Future for [`Async::writable`](crate::Async::writable).
-#[derive(Debug)]
 #[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct Writable(Ready);
+pub struct Writable<'a, T>(Ready<&'a crate::Async<T>, T>);
 
-impl Future for Writable {
+impl<T> Future for Writable<'_, T> {
     type Output = io::Result<()>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         ready!(Pin::new(&mut self.0).poll(cx))?;
-        log::trace!("writable: fd={}", self.0.source.raw);
+        log::trace!("writable: fd={}", self.0.handle.source.raw);
         Poll::Ready(Ok(()))
     }
 }
 
-#[derive(Debug)]
-struct Ready {
-    source: Arc<Source>,
+impl<T> fmt::Debug for Writable<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Writable").finish()
+    }
+}
+
+/// Future for [`Async::writable_owned`](crate::Async::writable_owned).
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct OwnedWritable<T>(Ready<Arc<crate::Async<T>>, T>);
+
+impl<T> Future for OwnedWritable<T> {
+    type Output = io::Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ready!(Pin::new(&mut self.0).poll(cx))?;
+        log::trace!("writable_owned: fd={}", self.0.handle.source.raw);
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<T> fmt::Debug for OwnedWritable<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OwnedWritable").finish()
+    }
+}
+
+struct Ready<S: Borrow<crate::Async<T>>, T> {
```

Comment:
Originally I wanted to use `AsRef<Source>`, as @Kestrer suggested in https://github.com/smol-rs/async-io/pull/64#issuecomment-846433555, but I could not implement it because we [had already implemented `AsRef<T>` for `Async<T>`](https://github.com/smol-rs/async-io/blob/e82546dbd93f1d44798de934d8c82f8dbd92c42f/src/lib.rs#L920).